### DISCUSSION TITLE
Experiment: `node-sea`

### DISCUSSION THREAD

**Discussion by arcanis** [nodejs/single-executable#40] [Show and tell]
Since the action item of the previous meeting was to build SEA experiments, I built mine: [node-sea](https://github.com/arcanis/node-sea).

It generates JS files that, when required or executed from the CLI, either re-export a module or run a binary. It uses the Yarn libraries in order to generate the dataset that can be stored, and applied at runtime as a layer on top of the filesystem.

Note that you'll currently need to build the `node` binary from sources (https://github.com/nodejs/node/pull/44537 is required), and some binaries such as Webpack will require https://github.com/nodejs/node/pull/44732 for workers to be supported.

```
npx node-sea -c 'npm add prettier' -b prettier -m prettier -o prettier
~/node/node ./prettier.cjs --help
```

**Comment by jviotti** [comment]
Amazing stuff @arcanis ! I'm making a note to take a deeper look into it soon. How are you embedding the generated JS code into the binary?

**Reply by arcanis** [reply]
It's stored as a base64 string, and unpacked on the fly with the native `Buffer.from(data, 'base64')` method

**Comment by jviotti** [comment]
Also, let me convert this issue into a GitHub Discussion on the "Show and Tell" category!

**Comment by jviotti** [comment]
@arcanis How would you feel about sending a PR to this repo including a document (maybe inside an `implementations/`) folder that describes in more detail what you built, the decisions you make (i.e. what did you use as a VFS, how did you taught Node.js about it, how did you inject the data on the binary, etc), the reasoning behind them, etc. Some sort of architecture review document.

I tried to do this myself while going through the `node-sea` code, but I think you are much better positioned to do this given you wrote the code!

Once the document is in place, we should add it to a future meeting agenda to discuss the results of the experiment and see what conclusions and decisions we can draw out of it.
